first attempts at writing the serializers for the new types PhylogeneticModel and GroupBasedPhylogeneticModel

mariusneu opened the pull request:
I am currently trying to serialize the new data types introduced in the algebraic statistics (sub)project on algebraic phylogenetics by @marinagarrote et al (cf. #3812). Following the documentation and writing the code as naive as possible, I got problems when testing the functions by trying to save an instance of GroupPhylogeneticModel, with the following error message: `Unsupported type 'GroupBasedPhylogeneticModel' for encoding.`

mariusneu committed:
first attempts at writing the serializers for the new types PhylogeneticModel and GroupBasedPhylogeneticModel

antonydellavecchia committed:
include serialization functions from AlgebraicStatistics

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,58 @@
+@register_serialization_type PhylogeneticModel uses_params 
+@register_serialization_type GroupBasedPhylogeneticModel uses_params 
```
if you set `use_params` the save mechanices will expect there to be a save_type_params function implemented.

mariusneu replied:
Removed that for now

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,58 @@
+
+function save_object(s::SerializerState, pm::PhylogeneticModel)
+    save_data_dict(s) do 
+        save_object(s, pm.graph, :tree) # already implemented 
```
does graph(pm) work?  in general you should avoid acessing properties and use functions, and if the functions don't exist you should considering writting them.

mariusneu replied:
Now I use functions throughout the code instead of properties.

mariusneu committed:
Removed used_params for both types, save_object uses functions instead of properties now

mariusneu committed:
Added functions to serialize edges of graphs

mariusneu committed:
Changed indentation from 4 to 2

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,41 @@
+function save_object(s::SerializerState, pm::GroupBasedPhylogeneticModel)
+  save_data_dict(s) do
+    save_object(s, phylogenetic_model(pm), :phylomodel) # details see above
+    save_object(s, fourier_ring(pm), :fourier_ring) # already implemented
```
this should probably be saved as a reference, the best thing to do for now is just
```suggestion
    save_typed_object(s, fourier_ring(pm), :fourier_ring) # already implemented
```

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,41 @@
+function load_object(s::DeserializerState, ::Type{<: PhylogeneticModel})
+  graph = load_object(s, Graph{Directed}, :tree)
+  n_states = load_object(s, Int64, :states)
+  prob_ring = load_object(s, QQMPolyRing, :probability_ring)
```
this would then become
```suggestion
  prob_ring = load_typed_object(s, :probability_ring)
```

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,41 @@
+  save_data_dict(s) do 
+    save_object(s, graph(pm), :tree) # already implemented 
+    save_object(s, number_states(pm), :states) # already implemented 
+    save_object(s, probability_ring(pm), :probability_ring) # already implemented
```
```suggestion
    save_typed_object(s, probability_ring(pm), :probability_ring) # already implemented
```

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,41 @@
+
+function load_object(s::DeserializerState, ::Type{<: GroupBasedPhylogeneticModel})
+  phylomodel = load_object(s, PhylogeneticModel, :phylomodel)
+  fourier_ring = load_object(s, QQMPolyRing, :fourier_ring)
```
```suggestion
  fourier_ring = load_typed_object(s, :fourier_ring)
```

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,41 @@
+    save_object(s, graph(pm), :tree) # already implemented 
+    save_object(s, number_states(pm), :states) # already implemented 
+    save_object(s, probability_ring(pm), :probability_ring) # already implemented
+    save_object(s, root_distribution(pm), :root_distribution) # needs to be implemented, or we change the entry type to QQFieldElem
```
you can using save_typed_object here if the type isn't always know, or if your using a vector any because not all entries are the same type you'll need to use a tuple

antonydellavecchia commented on src/Serialization/Combinatorics.jl:
```diff
@@ -19,6 +20,19 @@ function load_object(s::DeserializerState, g::Type{Graph{T}}) where T <: Union{D
 end
 
+function save_object(s::SerializerState, e::Edge)
+  save_data_dict(s) do 
```
this can be done using save_data_array instead, it'll make file sizes smaller

mariusneu committed:
Apply suggestions from code review: Save all polynomial rings as typed object

Co-authored-by: antonydellavecchia <antonydellavecchia@gmail.com>

mariusneu committed:
Changed tpye of root_distribution to Vector. Dirty workaround to allow loading

mariusneu committed:
Resolved conflicts

mariusneu committed:
Changed serialization of Edges to a more efficient way

antonydellavecchia committed:
saving loading working

mariusneu committed:
Some more changes in the serialization file as well as in the serialization of dicts to catch the case where the key is not serialized with params

antonydellavecchia commented on experimental/AlgebraicStatistics/src/serialization.jl:
```diff
@@ -0,0 +1,42 @@
+
+
+function load_object(s::DeserializerState, g::Type{PhylogeneticModel})
+  graph = load_object(s, Graph{Directed}, :tree)
```
here you would need to use 
```suggestion
  graph = load_typed_object(s, :tree)
```

antonydellavecchia committed:
jukes cantor model saving and loading

mariusneu committed:
Changed documentation, removed unneccessary comments

antonydellavecchia committed:
some progress on simplifying ref handling in IPC Serialization

antonydellavecchia committed:
some progress

antonydellavecchia committed:
moving things around

antonydellavecchia committed:
type params function

antonydellavecchia committed:
some progress on refactoring

antonydellavecchia committed:
some progress

mariusneu committed:
Wrote first mini test

mariusneu committed:
Wrote tests to verify serialization on object-identity level

antonydellavecchia committed:
more progress

mariusneu committed:
Changed type of root_distribution entries to QQ in all models. Made tests more elaborate. Changed testing for graphs so that false negatives are avoided.

mariusneu commented:
In commit 8723fc9, tests did not pass so I investigated it and found out that apparently object identity does not the same with polymake-derived objects? Therefore, the graph objects of the original model and the loaded one were not identical, although they were the same tree. I wrote a workaround using `adjacency_matrix(graph)`

antonydellavecchia committed:
almost completed uni variate polynomials

antonydellavecchia committed:
uni vairate polynomials serialization passing

antonydellavecchia committed:
MPolyRing tests passing

antonydellavecchia committed:
moving things around

antonydellavecchia committed:
grading needs containers

antonydellavecchia committed:
not sure where i left this

antonydellavecchia committed:
new load_type_params

antonydellavecchia committed:
bug with julia 1.11.1

antonydellavecchia committed:
vectors + polynomials working

antonydellavecchia committed:
some progress on tuples

antonydellavecchia committed:
reworking save_type_params

antonydellavecchia committed:
tidying

antonydellavecchia committed:
dicts working

antonydellavecchia committed:
containers completed, need some upgrades though

antonydellavecchia committed:
tidying

antonydellavecchia committed:
broke everything

antonydellavecchia committed:
fixed containers and polynomials

antonydellavecchia committed:
some issues with field embeddings

antonydellavecchia committed:
fields tests working now

antonydellavecchia committed:
algebras working

antonydellavecchia committed:
started on groups

antonydellavecchia committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
fix, inconsistencies from merge

antonydellavecchia committed:
crawling through groups

antonydellavecchia committed:
matrices working

antonydellavecchia committed:
MPolyMap Working

antonydellavecchia committed:
some progress on polyhedral objects

ThomasBreuer committed:
adjust GAP serialization to the refactoring

- introduce `install_GAP_type_params` that implements
  `type_params` methods based on GAP filters
- adjust code for `IsFreeGroup` and `IsSubgroupFpGroup`;
  now their tests work again

antonydellavecchia committed:
GAP object serialization working

antonydellavecchia committed:
groups working

antonydellavecchia committed:
removed false kwarg in docs

antonydellavecchia committed:
Lie Theory serialization working

antonydellavecchia committed:
stuck on regression in polyhedral

antonydellavecchia committed:
simplify type_params

antonydellavecchia committed:
fields working

antonydellavecchia committed:
back to polyhedral

antonydellavecchia committed:
graded ring working

antonydellavecchia committed:
sets not passing yet

antonydellavecchia committed:
started to change docs a bit

antonydellavecchia committed:
fix some set tests

antonydellavecchia committed:
TypeParams Struct

antonydellavecchia committed:
progress on polyhedral objects

antonydellavecchia committed:
progress on some polyhedral object types

antonydellavecchia committed:
more progress

HechtiDerLachs committed:
Fix for type_params of dictionaries to avoid infinite recursion.

benlorenz committed:
serialization: flatten some polymake object dicts

antonydellavecchia committed:
fix dict type_params

antonydellavecchia committed:
qqbar fix

antonydellavecchia committed:
second line fix for QQbar Field

antonydellavecchia committed:
proper fix

antonydellavecchia committed:
fixes for container tests

antonydellavecchia committed:
some progress with TypeParams Struct

antonydellavecchia committed:
Algebras and Fields done

antonydellavecchia committed:
groups working again

antonydellavecchia committed:
lie theory + MPolyMap

antonydellavecchia committed:
bunch more types working

antonydellavecchia committed:
tests passing for all types in src (except upgrade test)

antonydellavecchia committed:
cleaning up polymake dict functions

lgoettgens committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
removed redundant function

antonydellavecchia committed:
Merge remote-tracking branch 'origin/adv/ipc-serialize' into adv/ipc-serialize

antonydellavecchia committed:
Update src/Serialization/Groups.jl

Co-authored-by: Lars Göttgens <lars.goettgens@rwth-aachen.de>

lgoettgens committed:
Serialization refactor (Lie theory part) (#4563)

* Use `grading_group` for `MPolyDecRing`

* Fix some formatting

* Some progress

* Fix loading AbstractLieAlgebra

* LinearLieAlgebras working

* Modules working

* Attrs are handled centrally now

* Root system data working

* Module construction data works

* DirectSumLieAlgebras work

* oops

* Run formatter

* Reduce diff

antonydellavecchia committed:
remove dead code

antonydellavecchia committed:
remove obsolete code see pr #4501

antonydellavecchia committed:
can almost load example qsm

antonydellavecchia committed:
qsm saving and loading refactor

antonydellavecchia committed:
starting on upgrade

lgoettgens committed:
Merge branch 'master' into adv/ipc-serialize

antonydellavecchia committed:
fix loading vectors

HereAround committed:
[FTheoryTools] Update serialization of Weierstrass models

HereAround committed:
[FTheoryTools] Update serialization of GlobalTateModel

antonydellavecchia committed:
allow dict overrides

antonydellavecchia committed:
upgrade fqPolyrepFields

antonydellavecchia committed:
some upgrades

antonydellavecchia committed:
start on moving ftheory attributes to attribute framework

antonydellavecchia committed:
upgrade surfaces

antonydellavecchia committed:
upgrades for quad spaces

antonydellavecchia committed:
update enrique surfaces

antonydellavecchia committed:
Vector needs upgrade

antonydellavecchia committed:
issue with storing johnson solids after update

antonydellavecchia committed:
fix

antonydellavecchia committed:
comment out print statements

antonydellavecchia committed:
upgrade johnson solids

antonydellavecchia committed:
revert to using dev version on save

antonydellavecchia committed:
update catalan and archimedean solids

antonydellavecchia committed:
upgrades for all types in src

antonydellavecchia committed:
forgot to add file with ne extension

antonydellavecchia committed:
updates for loading IdealGens

antonydellavecchia committed:
improved Dict serialization coverage

antonydellavecchia committed:
catch missed upgrades

antonydellavecchia committed:
Merge remote-tracking branch 'origin/master' into adv/ipc-serialize

antonydellavecchia committed:
Merge remote-tracking branch 'origin/master' into adv/ipc-serialize

antonydellavecchia committed:
fixes

lgoettgens committed:
Serialization Refactor (Lie Theory again) (#4649)

* Update LieAlgebra serialization again

* Progress towards Lie algebra module construction data

* Move direct summands to type params

* Simplify standard module

* Changes to construction data again

* handles TypeParams with tuples

* fix for non refs types in tuple

* Remaining fix

---------

Co-authored-by: Antony Della Vecchia <vecchia@math.tu-berlin.de>

antonydellavecchia committed:
update versions

antonydellavecchia committed:
upgrades ZZLatwithisom and some lie algebra upgrades

mariusneu committed:
merged adv/ipc-serialize into master

fingolfin commented:
What the status of this PR? (Besides the fact that it has many merge conflicts now)

antonydellavecchia commented:
Serialization of phylogenetic trees will now be handle after the refactoring of the algebraic stats section.
